src/ecParsetree.ml

```diff
@@ -170,7 +122,37 @@ type glob_or_var =
   | GVglob of pmsymbol located * pqsymbol list
   | GVvar  of pqsymbol
 
-type pformula  = pformula_r located
+(* -------------------------------------------------------------------- *)
+type plvalue_r =
+  | PLvSymbol of pqsymbol
+  | PLvTuple  of pqsymbol list
+  | PLvMap    of pqsymbol * ptyannot option * pexpr list
+
+and plvalue = plvalue_r located
+
+and pinstr_r =
+  | PSident  of psymbol
+  | PSasgn   of plvalue * pexpr
+  | PSrnd    of plvalue * pexpr
+  | PScall   of plvalue option * pgamepath * (pexpr list) located
+  | PSif     of pscond * pscond list * pstmt
+  | PSwhile  of pscond
+  | PSmatch  of pexpr * psmatch
+  | PSassert of pexpr
+
+and psmatch = [
+  | `Full of (ppattern * pstmt) list
+  | `If   of (ppattern * pstmt) * pstmt option
+]
+
+and pscond = pexpr * pstmt
+and pinstr = pinstr_r located
+and pstmt  = pinstr list
+
+and pexpr   = pexpr_r located
+and pexpr_r = Expr of pformula
```

Comment:
It is really useful to have the Expr constructor ?

Reply:
Yes. Internally, expressions & formulas deserve different purposes and the high-level tactic API takes parse trees. This forbids using one in place of the other.